File: java/MaximumProductSubarray.java
/*
 * Filename: MaximumProductSubarray.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 20 Jun 2015 04:47:26 PM CST
 */

public class Solution {
	public int maxProduct(int nums[]) {
		if (nums == null || nums.length == 0) {
			return 0;
		}

		int res = nums[0];
		int max = nums[0];
		int min = nums[0];

		for (int i = 1; i < nums.length; i++) {
			int a = nums[i] * max;
			int b = nums[i] * min;
			max = Math.max(Math.max(a, b), nums[i]);
			min = Math.min(Math.min(a, b), nums[i]);
			res = Math.max(res, max);
		}

		return res;
	}
}

File: cpp/maximum_product_subarray.cc
/*
 * Filename: maximum_product_subarray.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 07 Aug 2015 06:46:50 PM EDT
 */

#include <climits>
#include <iostream>

using namespace std;

class Solution {
	public:
		int maxProduct(vector<int>& nums) {
			if (nums.empty())
			{
				return 0;
			}

			int maxi = nums[0];
			int mini = nums[0];
			int res = nums[0];

			for (vector<int>::size_type i = 1; i < nums.size(); i++)
			{
				int a = maxi * nums[i];
				int b = mini * nums[i];
				maxi = max(nums[i], max(a, b));
				mini = min(nums[i], min(a, b));
				res = max(res, maxi);
			}

			return res;
		}
};
